src/coreclr/jit/gentree.h

```diff
@@ -350,6 +351,341 @@ struct Statement;
+    GTF_INX_STRING_LAYOUT       = 0x40000000, // GT_INDEX -- this uses the special string array layout
+
+    GTF_IND_TGT_NOT_HEAP        = 0x80000000, // GT_IND   -- the target is not on the heap
+    GTF_IND_VOLATILE            = 0x40000000, // GT_IND   -- the load or store must use volatile sematics (this is a nop on X86)
+    GTF_IND_NONFAULTING         = 0x20000000, // Operations for which OperIsIndir() is true  -- An indir that cannot fault.
+                                              // Same as GTF_ARRLEN_NONFAULTING.
+    GTF_IND_TGTANYWHERE         = 0x10000000, // GT_IND   -- the target could be anywhere
+    GTF_IND_TLS_REF             = 0x08000000, // GT_IND   -- the target is accessed via TLS
+    GTF_IND_ASG_LHS             = 0x04000000, // GT_IND   -- this GT_IND node is (the effective val) of the LHS of an
+                                              //             assignment; don't evaluate it independently.
+    GTF_IND_REQ_ADDR_IN_REG     = GTF_IND_ASG_LHS, // GT_IND  -- requires its addr operand to be evaluated
+                                              // into a register. This flag is useful in cases where it
+                                              // is required to generate register indirect addressing mode.
+                                              // One such case is virtual stub calls on xarch.  This is only
+                                              // valid in the backend, where GTF_IND_ASG_LHS is not necessary
+                                              // (all such indirections will be lowered to GT_STOREIND).
+    GTF_IND_UNALIGNED           = 0x02000000, // GT_IND   -- the load or store is unaligned (we assume worst case
+                                              //             alignment of 1 byte)
+    GTF_IND_INVARIANT           = 0x01000000, // GT_IND   -- the target is invariant (a prejit indirection)
+    GTF_IND_ARR_INDEX           = 0x00800000, // GT_IND   -- the indirection represents an (SZ) array index
+    GTF_IND_NONNULL             = 0x00400000, // GT_IND   -- the indirection never returns null (zero)
+
+    GTF_IND_FLAGS = GTF_IND_VOLATILE | GTF_IND_TGTANYWHERE | GTF_IND_NONFAULTING | GTF_IND_TLS_REF | \
+                    GTF_IND_UNALIGNED | GTF_IND_INVARIANT | GTF_IND_NONNULL | GTF_IND_ARR_INDEX | GTF_IND_TGT_NOT_HEAP,
+
+    GTF_CLS_VAR_VOLATILE        = 0x40000000, // GT_FIELD/GT_CLS_VAR -- same as GTF_IND_VOLATILE
+    GTF_CLS_VAR_INITCLASS       = 0x20000000, // GT_FIELD/GT_CLS_VAR -- same as GTF_FLD_INITCLASS
+    GTF_CLS_VAR_ASG_LHS         = 0x04000000, // GT_CLS_VAR   -- this GT_CLS_VAR node is (the effective val) of the LHS
+                                              //                 of an assignment; don't evaluate it independently.
+
+    GTF_ADDRMODE_NO_CSE         = 0x80000000, // GT_ADD/GT_MUL/GT_LSH -- Do not CSE this node only, forms complex
```

Comment:
It was odd earlier, but now it would be more obvious to have same value (e.g. `0x80000000`) to multiple enum. Do we need it that way? Can we have a single value / flag to represent that they mean same thing?

Reply:
Yes, it makes sense to have different names for the same value, because there are the "common" bits (the low bits) and the node-type-specific bits (the high bits). The specific bits have different names because they mean different things to different node types. It's kind of a class hierarchy but in bits.

Reply:
Would be interesting to figure out how to make this into a tagged union or similar, so we have some guarantee we're not inadvertently mixing bits (which has happened, and yes I know there are some cases where we deliberately do this).

Or perhaps some special build where all these flags have distinct values and we expand the underlying storage somehow (problematic with the use of `gtFlags` the way we do, we'd need to make that into a datatype that has overloads for all the common tricks we play).

Reply:
>  we're not inadvertently mixing bits

Yes, that's the part I am worried about.